Artists support

imLinguin opened the pull request:
- [x] Search for Artists
- [x] Get data about Artist
- [x] List songs (artist object contains playlistId with all of the songs)
- [x] List carousel albums
- [ ] List albums from SEE ALL (requires another request to YouTube)
- [x] Show similar Artists

imLinguin committed:
Artists support with basic albums detection, MusicVideo now contains multiple artists

imLinguin committed:
Parsing suggestedArtists

imLinguin commented:
In terms of SEE ALL albums button, I couldn't get it working, it always returned page without `gridRenderer` Object that contains all the albums. If someone finds what causes it please let me know. I tested it with every potentially valid header I could.

imLinguin commented:
I looked more into that. Looks like X-Goog-Visitor-Id header is required. We need to find a way to get it.

baptisteArno commented:
I've just tried a bit. This is working fine:

```
curl --location --request POST 'https://music.youtube.com/youtubei/v1/browse?key=AIzaSyC9XL3ZjWddXya6X74dJoCTL-WEYFDNX30' \
--header 'origin: https://music.youtube.com' \
--header 'Content-Type: application/json' \
--data-raw '{
    "context": {
        "client": {
            "clientName": "WEB_REMIX",
            "clientVersion": "0.1",
            "hl": "fr",
            "gl": "FR"
        }
    },
   "browseId":"VLPLy9-FeudRYih9kwdsL_1xOq2DzkQnAwuG",
}'
```

`browseId` is the "See all" id

imLinguin commented:
> `"browseId":"VLPLy9-FeudRYih9kwdsL_1xOq2DzkQnAwuG",`

This is a playlist ID which contains all the songs. I've already implemented it as `songsPlaylistId?: string;` (better naming required). I thought that we should leave fetching all the songs to the user using `listMusicsFromPlaylist`.
I already implemented parsing albums from carousel (`albums?:AlbumPreview[];`). As SEE ALL I was referring to the **Albums** one.
 Since there might be a lot more of them than in the carousel.
![obraz](https://user-images.githubusercontent.com/62100117/136417187-d90e566d-f4ca-43e4-9e0c-b62e71f9912c.png)

Which provides browseId and the params. **Note: browseId is the artist id. params identifies what we want.**
```json
"moreContentButton": {
    "buttonRenderer":{},
    "navigationEndpoint": {
      "clickTrackingParams": "COYEEOvLBSITCO3FxO7NuPMCFYWP3godGqYNFQ==",
      "browseEndpoint": {
        "browseId": "UClmXPfaYhXOYsNn_QUyheWQ",
        "params": "6gPbAUNxRUJDb0VCQ25JQUFIQnNBQUZRVEFBQlVFd0FBUUJHUlcxMWMybGpYMlJsZEdGcGJGOWhjblJwYzNRQUFRQUJRd0FBQUFFQUFRQUFBUUVFQVo5emZRaVhBeG9ZVlVOc2JWaFFabUZaYUZoUFdYTk9ibDlSVlhsb1pWZFJnZ0VZVlVOc2JWaFFabUZaYUZoUFdYTk9ibDlSVlhsb1pWZFJBQUVRdjd5djdzMjQ4d0lhQW1WcUdBQXFEMkZ5ZEdsemRGOXlaV3hsWVhObGN6Q3gxTkRsbF9ISjhuQQ%3D%3D"
      }
    }
  }
```

baptisteArno commented:
Oh yeah my bad!

Indeed, it's a tricky one...

Let's ignore this for the moment. Let me know when the PR ready to get reviewed 👍

imLinguin committed:
Album Types, Singles and Albums separated in Artist object

imLinguin commented:
# Summary of this PR

- Detecting multiple artists in MusicVideo
- Ability to search for artists
- Get more data about artists (songs, albums, singles, suggested artists, description, thumbnails)
- Albums now have types, which can be easily distinguished using `AlbumType` enum
- Added `PageType` enum which allows to detect type of the item *useful in searching without filters - can be added soon*

imLinguin committed:
Better search handling (when query was misspeled no results were shown)

baptisteArno commented:
Nice job man!
